### PR TITLE
Timetable search updates 

### PR DESCRIPTION
**Pull request by david-yz-liu**
1. Users may select a department, and those courses load in the timetable.
2. Improved the search interface.
3. Enable search by time.

I'll leave this up for review for a while. @Ian-Stewart-Binks :)

### CODE REVIEW

**Comment by Ian-Stewart-Binks** [comment]
Thanks :)

File: public/js/search/timetable.js [hunk 1]
```diff
@@ -1,10 +1,86 @@
+var filterCourse = function(inst, time, lec) {
```
Comment (Ian-Stewart-Binks) [inline public/js/search/timetable.js]: These functions are missing `use strict;` statements.

File: public/js/search/timetable.js [hunk 2]
```diff
@@ -1,10 +1,86 @@
...
+    return lec.instructor.indexOf(inst) > -1 && 
+           (time.length < 2 || hasTime(time, lec.time));
+};
+
+var hasTime = function(timeStr, times) {
+    var time = ["MTWRF".indexOf(timeStr[0]) + "",
```
Comment (Ian-Stewart-Binks) [inline public/js/search/timetable.js]: `"`

File: public/js/search/timetable.js (hunk 1 above)
Comment (Ian-Stewart-Binks) [inline public/js/search/timetable.js]: Missing space between `function` and `(`.

File: public/js/search/timetable.js [hunk 3]
```diff
@@ -1,10 +1,86 @@
+var filterCourse = function(inst, time, lec) {
+    return lec.instructor.indexOf(inst) > -1 && 
+           (time.length < 2 || hasTime(time, lec.time));
+};
+
```
Comment (Ian-Stewart-Binks) [inline public/js/search/timetable.js]: Should be two blank lines.

File: hs/Database/CourseQueries.hs [hunk 4]
```diff
@@ -124,6 +125,16 @@ courseInfo = do
             Just $ Tables.Session (map buildLecture lectures)
                                   (map buildTutorial tutorials)
 
+-- | Return a list of all departments
+deptList :: IO Response
+deptList = do
+    depts <- runSqlite dbStr $ do
+        courses :: [Entity Courses] <- selectList [] []
+        return $ sort . nub $ map f courses
+    return $ createJSONResponse $ encodeJSON $ Aeson.toJSON depts
```
Comment (Ian-Stewart-Binks) [inline hs/Database/CourseQueries.hs]: We seem to use `encodeJSON` along with `createJSONResponse` in both uses of `createJSON` response. Should we absorb encodeJSON in createJSONResponse? We also use Aeson.toJSON in both as well.

**Comment by david-yz-liu** [comment]
Thanks, @Ian-Stewart-Binks

**Pull request merged**